bpf: daemon: Use bpftool for BPF feature probes

vadorovsky opened the pull request:
bpftool and libbpf have feature checks, which have similar goals to BPF
feature checks in Cilium. This change makes use of them and drops the
usage of run_probes.sh script which is not needed anymore.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8094)
<!-- Reviewable:end -->

vadorovsky committed:
bpf: daemon: Use bpftool for BPF feature probes

bpftool and libbpf have feature checks, which have similar goals to BPF
feature checks in Cilium. This change makes use of them and drops the
usage of run_probes.sh script which is not needed anymore.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
bpf: Use bpftool feature header defines

Some BPF features checks are performed by bpftool now and names of
macros are different.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
bpf: probes: Remove old style probes

BPF feature probes are now mostly handled by bpftool. There are only few
old style probes in Cilium which still are not handled by bpftool, which
are the only not removed by this change.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
bpf: lib: Use both new and legacy BPF feature checks

The old style BPF feature checks performed by Cilium save mactos in
bpf_features_legacy.h now. bpf_features.h is used by bpftool.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
bpf: Change name of the old style checks header file

bpf_features.h name is going to be used by new style bpftool checks, so
the old one is renamed to bpf_features_legacy.h.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

joestringer commented on bpf/lib/common.h:
```diff
@@ -19,6 +19,7 @@
 #define __LIB_COMMON_H_
 
 #include <bpf_features.h>
+#include <bpf_features_legacy.h>
```
I think that this commit and the previous need to be the same commit, or something will break?

vadorovsky replied:
It should be fine.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -286,21 +287,36 @@ func checkMinRequirements() {
 	}
+
+	// Old Cilium BPF feature probes.
+	// TODO(mrostecki): Remove them completely and move missing checks to
```
Will you address this TODO In this PR or a subsequent PR?

vadorovsky replied:
In a subsequent PR. We have some checks that are not implemented in bpftool, so I need to first send patches.

By the way, since we are already talking about that. Is it fine if I send the missing probes to the bpf list and add `Signed-off-by` of all authors who modified probe files, and the info that it comes from Cilium?

joestringer replied:
I would probably only add `Signed-off-by` if I had asked the relevant authors if that was okay (which you could likely do via Slack). Otherwise there is `Authored-by` / `Co-authored-by` and/or just a note in the commit message pointing to the origin, `bpf/` directory license (which is GPLv2, hence compatible), and attributing the code to the relevant Cilium authors. It would also be good etiquette to CC the relevant people on the mail.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -286,21 +287,36 @@ func checkMinRequirements() {
+		log.WithError(err).Fatal("BPF check: NOT OK.")
+	}
+	if err := probeManager.SystemConfigProbes(); err != nil {
+		log.WithError(err).Fatal("BPF check: NOT OK.")
```
I'm pretty sure this will fail in many environments, eg I have an Ubuntu environment happily running current Cilium and here's my `bpftool -j feature` output:

```
$ k -n kube-system exec $(get_cilium_pod) -- bpftool -j feature | jq '.system_config'
{
  "unprivileged_bpf_disabled": 1,
  "bpf_jit_enable": 1,
  "bpf_jit_harden": 0,
  "bpf_jit_kallsyms": 0,
  "bpf_jit_limit": 264241152,
  "CONFIG_BPF": null,
  "CONFIG_BPF_SYSCALL": null,
  "CONFIG_HAVE_EBPF_JIT": null,
  "CONFIG_BPF_JIT": null,
  "CONFIG_BPF_JIT_ALWAYS_ON": null,
  "CONFIG_CGROUPS": null,
  "CONFIG_CGROUP_BPF": null,
  "CONFIG_CGROUP_NET_CLASSID": null,
  "CONFIG_SOCK_CGROUP_DATA": null,
  "CONFIG_BPF_EVENTS": null,
  "CONFIG_KPROBE_EVENTS": null,
  "CONFIG_UPROBE_EVENTS": null,
  "CONFIG_TRACING": null,
  "CONFIG_FTRACE_SYSCALLS": null,
  "CONFIG_FUNCTION_ERROR_INJECTION": null,
  "CONFIG_BPF_KPROBE_OVERRIDE": null,
  "CONFIG_NET": null,
  "CONFIG_XDP_SOCKETS": null,
  "CONFIG_LWTUNNEL_BPF": null,
  "CONFIG_NET_ACT_BPF": null,
  "CONFIG_NET_CLS_BPF": null,
  "CONFIG_NET_CLS_ACT": null,
  "CONFIG_NET_SCH_INGRESS": null,
  "CONFIG_XFRM": null,
  "CONFIG_IP_ROUTE_CLASSID": null,
  "CONFIG_IPV6_SEG6_BPF": null,
  "CONFIG_BPF_LIRC_MODE2": null,
  "CONFIG_BPF_STREAM_PARSER": null,
  "CONFIG_NETFILTER_XT_MATCH_BPF": null,
  "CONFIG_BPFILTER": null,
  "CONFIG_BPFILTER_UMH": null,
  "CONFIG_TEST_BPF": null
}
```

Looks like in this case, we just can't detect the kernel config, so we silently allow Cilium to continue running.

vadorovsky replied:
OK. Although it's very sad and weird that some distros don't provide kernel config. Are you sure that there is no workaround for Ubuntu.

Yeah, sorry for being focused only on RPM distros. ;)

joestringer replied:
For Ubuntu specifically, I can see a `/boot/config-$(uname -r)` file which apparently is not picked up / parsed by bpftool.

However, I don't think it's just Ubuntu, AFAIK CoreOS hides the config completely; I wouldn't be surprised if some of the cloud distros skip it as well. I'm all for encouraging distros to provide these configs, but unfortunately I don't think that there's a single isolated distro we can just pressure to fix this, there's a bunch of them.

joestringer commented on pkg/bpf/probes/probes.go:
```diff
@@ -0,0 +1,241 @@
+// features are available in the kernel.
+func CreateHeadersFile() error {
+	globalsDir := option.Config.GetGlobalsDir()
+	if err := os.MkdirAll(globalsDir, defaults.StateDirRights); err != nil {
```
Can `err` be `os.IsExist()` in this case? Then we won't want to `Fatal()` out.

vadorovsky replied:
You are right.

joestringer commented on pkg/bpf/probes/probes.go:
```diff
@@ -0,0 +1,241 @@
+func CreateHeadersFile() error {
+	globalsDir := option.Config.GetGlobalsDir()
+	if err := os.MkdirAll(globalsDir, defaults.StateDirRights); err != nil {
+		log.WithError(err).WithField(logfields.Path, globalsDir).Fatal(
```
Should this be more consistent with the rest of the function, ie. return an error and let the caller decide to `Fatal()`?

vadorovsky replied:
True.

joestringer commented on pkg/bpf/probes/probes.go:
```diff
@@ -0,0 +1,241 @@
+
+	go io.Copy(writer, stdoutPipe)
+	if err := cmd.Wait(); err != nil {
+		return fmt.Errorf("bpftool did not run successfully: %s", err)
```
In this case I think it would be useful to see the `stderr` of the process?

vadorovsky replied:
Right

joestringer commented on bpf/bpf_features_legacy.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef BPF_FEATURES_LEGACY_H_
+#define BPF_FEATURES_LEGACY_H_
+
+#define HAVE_SKB_CHANGE_TAIL
```
Should be able to switch this one to the new format too.

vadorovsky replied:
Good point

vadorovsky replied:
To be honest, I don't even see skb_change_tail used anywhere and the check for it is commented out:

https://github.com/cilium/cilium/blob/master/bpf/run_probes.sh#L130

So I will just remove it.

joestringer replied:
I think it's commented out because it became part of the regular `probe_run_ll()` case:

```
# grep CHANGE_TAIL /var/run/cilium/state/globals/bpf_features.h 
#define HAVE_SKB_CHANGE_TAIL
# git grep change_tail
bpf/include/bpf/api.h:static int BPF_FUNC(skb_change_tail, struct __sk_buff *skb, uint32_t nlen,
bpf/include/linux/bpf.h: * int bpf_skb_change_tail(struct sk_buff *skb, u32 len, u64 flags)
bpf/include/linux/bpf.h:        FN(skb_change_tail),            \
bpf/lib/icmp6.h:                if (skb_change_tail(skb, skb->len + trimlen, 0) < 0)
bpf/lib/icmp6.h:                if (skb_change_tail(skb, skb->len + trimlen, 0) < 0)
bpf/probes/raw_change_tail.t: * 5293efe62df8 ("bpf: add bpf_skb_change_tail helper")
bpf/probes/raw_change_tail.t:                   BPF_EMIT_CALL(BPF_FUNC_skb_change_tail),
bpf/probes/raw_change_tail.t:           .warn = "Kernel does not support bpf_skb_change_tail() helper. "
bpf/run_probes.sh:#probe_run_tc "skb_change_tail.c" "HAVE_SKB_CHANGE_TAIL"
```

joestringer reviewed (changes requested):
Sorry about the long delay on review. I think it looked like upstream changes were necessary so I put off reviewing.

Looks pretty good overall, I think it would be useful to retain successful compile/run for each commit in case we need to bisect an issue in the PR (kinda a pain, but the `bindata` update in the last commit means that earlier commits aren't buildable).

Please double-check the behaviour of the new and old versions with respect to the handling of each kind of detection. For instance, the kernel config is a bit special because kernel config is not always available. So failing out like the current version does will likely break setups like that. I haven't tried to run this PR, but posted some output of one of my environments to assist.

I have a few further comments below.

vadorovsky committed:
cmd: Initialize supported map types in BPF commands

The information about supported BPF map types is now retrieved from
bpftool checks and the structure needs to be initialized by each command
which is going to use it later.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
docker: Use bpftool from bpf-next instead of Linus' tree

After this change, bpftool in the runtime image is build from bpf-next
kernel tree, not the mainline tree.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

joestringer reviewed (approved):
LGTM, previous comments appear to have been addressed and I tested this with microk8s, both running the daemon and running some CLI tools. :+1:

aanm reviewed (commented):
Overall looks good to me but I'll defer to @joestringer
